keps/sig-node/0000-apparmor-ga/README.md

```diff
@@ -0,0 +1,682 @@
+
+- Defining any standard "Kubernetes branded" AppArmor profiles
+- Formally specifying the AppArmor profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside of the node
+- Changing the semantics around AppArmor support
+- Windows support
+
+## Proposal
+
+AppArmor is not available on every Linux distribution. Beside this, container
+runtimes have AppArmor as compile-time feature which may be disabled as well.
+With the GA API we do not change the error handling and behave exactly the same
+as the current error propagation paths.
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with.
+There are no changes planned to the current `PodSecurityPolicy` API as part of
+this KEP.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable, except in `PodTemplates`.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container. If AppArmor options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
+    // The `Type` in this struct has to be set to `AppArmorProfileLocalhost`.
+    // +optional
+    LocalhostProfile *string
+}
+
+type AppArmorProfileType string
+
+const (
+    AppArmorProfileTypeUnconfined     AppArmorProfileType = "Unconfined"
+    AppArmorProfileTypeRuntimeDefault AppArmorProfileType = "RuntimeDefault"
+    AppArmorProfileTypeLocalhost      AppArmorProfileType = "Localhost"
+)
+```
+
+This API makes the options more explicit and leaves room for new profile sources
+to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles) and for future extensions, such as defining the behavior when a
+profile cannot be set.
+
+##### Localhost Profile
+
+This KEP proposes we GA LocalhostProfile as the only source of user-defined
+profiles at this point. User-defined profiles are essential for users to realize
+the full benefits out of AppArmor, allowing them to decrease their attack
+surface based on their own workloads.
+
+###### Updating AppArmor profiles
+
+AppArmor profiles are applied at container creation time. The underlying
+container runtime loads the profile referenced by its name via
+`apparmor_parser` on container creation. Therefore, updating the profiles
+content will only apply if the workload gets restarted.
```

Comment:
IIRC, the runtime just uses profiles that have already been loaded by `apparmor_parser`, hence the need for the "apparmor-loader" image to load them: https://github.com/kubernetes/kubernetes/blob/69ae3144428e84069be928a8c70a356cb9bcb4b9/test/images/apparmor-loader/loader.go#L192

Reply:
Yes, I rephrased this paragraph to mention that the runtime just references already loaded profiles. I also added a note that the profile have to be reloaded manually if changed on disk.